### PR TITLE
Support training of BatchNormalization layer

### PR DESCRIPTION
**Pull request by caisq**
Fixes: https://github.com/tensorflow/tfjs/issues/20

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/139)

### CODE REVIEW

**Comment by nsthorat** [comment]
:lgtm_strong:

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[src/backend/tfjs_backend.ts, line 448 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/139#-LAhrxavDM9EtxMiza2g:-LAhrxavDM9EtxMiza2h:b-6mlz1t) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d3ca5fc289af13944850772073a431598b59344e/src/backend/tfjs_backend.ts#L448)):*
> ```XML
>     x: Tensor, gamma: Tensor, beta: Tensor, reductionAxes: number[],
>     epsilon = 1e-3): [Tensor, Tensor, Tensor] {
>   const meanAndVariance = tfc.moments(x, reductionAxes);
> ```

the body of these methods should contain a tf.tidy() (ideally you shouldn't rely on a parent tidy() to clean up intermediates of a function)

---

*[src/layers/normalization.ts, line 206 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/139#-LAhsfnN3h64trBCfSPi:-LAhsfnN3h64trBCfSPj:bo8vx14) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d3ca5fc289af13944850772073a431598b59344e/src/layers/normalization.ts#L206)):*
> ```XML
> 
>   // tslint:disable-next-line:no-any
>   call(inputs: Tensor|Tensor[], kwargs: any): Tensor|Tensor[] {
> ```

this method should have a tidy() in the body since there are a lot of intermediates here. 

---

*[src/layers/normalization_test.ts, line 278 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/139#-LAhspCq53R4_mxz9lwb:-LAhspCq53R4_mxz9lwc:bc76y89) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d3ca5fc289af13944850772073a431598b59344e/src/layers/normalization_test.ts#L278)):*
> ```XML
>       const movingVarianceValue = layer1.getWeights()[3];
>       expectTensorsClose(movingVarianceValue, [20.270758, 12.269142]);
>       done();
> ```

If you have an error here it will just say "async timer never fired". I would do this with model.fit.then() and then call "done" in the body of the then.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/139#-:-LAmt4HA99kFK3qyhUQr:b-jbsf8z)*

**Comment by bileschi** [comment]
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/backend/tfjs_backend.ts, line 491 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/139#-LAmxMHU2ucQ6z4aIsif:-LAmxMHU2ucQ6z4aIsig:b-6c9ke) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d3ca5fc289af13944850772073a431598b59344e/src/backend/tfjs_backend.ts#L491)):*
> ```
>  * Batch normalization for use in training (not inference).
>  *
> ```
Please describe here and above the normalization that takes place.  I.e., 0-mean and unit var.  Thanks.

---

*[src/backend/tfjs_backend_test.ts, line 614 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/139#-LAmxwy214n8Vg0NDgUL:-LAmxwy214n8Vg0NDgUM:b-dpavi6) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d3ca5fc289af13944850772073a431598b59344e/src/backend/tfjs_backend_test.ts#L614)):*
> ```
>  
> describeMathCPUAndGPU('normalizeBatchInTraining', () => {
> ```
Please describe how the golden values were determined.  Thanks.

---

*[src/layers/normalization.ts, line 262 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/139#-LAmwHi18jyZYsyIjpZA:-LAmwHi18jyZYsyIjpZB:bppugv1) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d3ca5fc289af13944850772073a431598b59344e/src/layers/normalization.ts#L262)):*
<details><summary><i alt="summary">Quoted 8 lines of code…</i></summary>

> ```
>       const newMovingMean = movingAverage(
>           this.movingMean.read(), mean, this.momentum, this.stepCount);
>       this.movingMean.write(newMovingMean);
>       const newMovingVariance = movingAverage(
>           this.movingVariance.read(), varianceDebiased, this.momentum,
>           this.stepCount);
>       this.movingVariance.write(newMovingVariance);
>     };
> ```
</details>

Why the immediately executed function pattern here?  Is it for variable scope management?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/139)*

**Comment by caisq** [comment]
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/backend/tfjs_backend.ts, line 448 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/139#-LAhrxavDM9EtxMiza2g:-LAmxV8y4dMedkzYO78G:b-fnpyd6) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d3ca5fc289af13944850772073a431598b59344e/src/backend/tfjs_backend.ts#L448)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

the body of these methods should contain a tf.tidy() (ideally you shouldn't rely on a parent tidy() to clean up intermediates of a function)
</blockquote></details>

Done. Sorry I should remember this since you've raised similar comments before.

---

*[src/backend/tfjs_backend.ts, line 491 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/139#-LAmxMHU2ucQ6z4aIsif:-LAn-oPF7aJbaUSRDil6:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d3ca5fc289af13944850772073a431598b59344e/src/backend/tfjs_backend.ts#L491)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
>  * Batch normalization for use in training (not inference).
>  *
> ```
Please describe here and above the normalization that takes place.  I.e., 0-mean and unit var.  Thanks.
</blockquote></details>

Done.

---

*[src/backend/tfjs_backend_test.ts, line 614 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/139#-LAmxwy214n8Vg0NDgUL:-LAn1hlaBdot_AFXvg1d:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d3ca5fc289af13944850772073a431598b59344e/src/backend/tfjs_backend_test.ts#L614)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
>  
> describeMathCPUAndGPU('normalizeBatchInTraining', () => {
> ```
Please describe how the golden values were determined.  Thanks.
</blockquote></details>

Done.

---

*[src/layers/normalization.ts, line 206 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/139#-LAhsfnN3h64trBCfSPi:-LAmxZEnCnnvzy_agmbv:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d3ca5fc289af13944850772073a431598b59344e/src/layers/normalization.ts#L206)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

this method should have a tidy() in the body since there are a lot of intermediates here. 
</blockquote></details>

Done.

---

*[src/layers/normalization.ts, line 262 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/139#-LAmwHi18jyZYsyIjpZA:-LAn1kKyC_ajSREYwh6A:b-puqsr0) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d3ca5fc289af13944850772073a431598b59344e/src/layers/normalization.ts#L262)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
>       const newMovingMean = movingAverage(
>           this.movingMean.read(), mean, this.momentum, this.stepCount);
>       this.movingMean.write(newMovingMean);
>       const newMovingVariance = movingAverage(
>           this.movingVariance.read(), varianceDebiased, this.momentum,
>           this.stepCount);
>       this.movingVariance.write(newMovingVariance);
>     };
> ```
Why the immediately executed function pattern here?  Is it for variable scope management?
</blockquote></details>

This has to do with the fact that in PyKeras, these updates to movingMean and movingAverage are done as a deferred Graph. I want to structurally call this out to make it clear. I added a Porting Note here.

---

*[src/layers/normalization_test.ts, line 278 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/139#-LAhspCq53R4_mxz9lwb:-LAmyezR0WOqWp3szC-J:b-snh328) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d3ca5fc289af13944850772073a431598b59344e/src/layers/normalization_test.ts#L278)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

If you have an error here it will just say "async timer never fired". I would do this with model.fit.then() and then call "done" in the body of the then.
</blockquote></details>

Done here and elsewhere.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/139)*
